### PR TITLE
Ensure Range streams and RangeToolLink respect subcoordinate axis range

### PR DESCRIPTION
**Pull request by philippjfr**
This PR handles the previously poorly defined behavior of Range streams and RangeToolLink attached to a subcoordinate axis. Specifically both now pick up the range of the outer subcoordinate axis range rather than one of the inner subcoordinate ranges. The latter case is still something that will be needed in certain cases so that is still a problem to be solved.

Fixes https://github.com/holoviz/holoviews/issues/6010 and  https://github.com/holoviz/holoviews/issues/6136

### CODE REVIEW

**Comment by philippjfr** [comment]
@droumis please test when you get a chance.

**Comment by ahuang11** [comment]
It seems to work. Awesome!
image

image

**Comment by philippjfr** [comment]
@ahuang11 Remind me what the issue we encountered in the CZI call yesterday was?

**Comment by ahuang11** [comment]
After zooming in, then zooming out, the curves end up offset from the yticks.

**Comment by philippjfr** [comment]
Ah, yes, thanks!

**Comment by philippjfr** [comment]
@droumis @ahuang11 Should be fixed now.

**Comment by ahuang11** [comment]
Are the changes pushed somewhere?

**Comment by ahuang11** [comment]
Works for me!

**Review by ahuang11 — APPROVED** [review]

**Comment by philippjfr** [comment]
@hoxbro This should be ready to review again.

File: holoviews/plotting/plot.py [hunk 1]
```diff
@@ -979,10 +979,21 @@ def _construct_callbacks(self):
...
+                if (src is not source and (src._plot_id is None or src._plot_id != source._plot_id)):
+                    continue
+                for stream in src_streams:
+                    # Skip if Stream.source is an overlay but the plot isn't
+                    # or if the source is an element but the plot isn't
+                    src_el = stream.source.last if isinstance(stream.source, HoloMap) else stream.source
+                    if ((isinstance(src_el, CompositeOverlay) and
+                        not isinstance(self, GenericOverlayPlot)) or
+                        (isinstance(src_el, Element) and
+                         isinstance(self, GenericOverlayPlot))):
```
Comment (hoxbro) [inline holoviews/plotting/plot.py:994]:
```suggestion
                    if (
                        (isinstance(src_el, CompositeOverlay) and not isinstance(self, GenericOverlayPlot)) 
                        or (isinstance(src_el, Element) and isinstance(self, GenericOverlayPlot))
                    ):
```

Easier for me to read.

File: holoviews/tests/plotting/bokeh/test_callbacks.py [hunk 2]
```diff
@@ -487,3 +487,42 @@ def test_rangexy_multi_yaxes():
...
+            Curve(-np.arange(100).cumsum(), vdims='y2', label='B').opts(subcoordinate_y=True)
+        )
+
+    stream = RangeXY()
+    dmap = DynamicMap(cb, streams=[stream])
+    plot = bokeh_server_renderer.get_plot(dmap)
+
+    p1, p2 = plot.subplots.values()
+
+    assert not (p1.callbacks or p2.callbacks)
```
Comment (hoxbro) [inline holoviews/tests/plotting/bokeh/test_callbacks.py:525]:
```suggestion
    assert not p1.callbacks 
    assert not p2.callbacks
```

It's easier for me to read, and if one of them starts failing, it's also easier to see which.

File: holoviews/tests/plotting/bokeh/test_callbacks.py [hunk 3]
```diff
@@ -487,3 +487,42 @@ def test_rangexy_multi_yaxes():
...
+    c2 = Curve(-np.arange(100).cumsum(), vdims='y2', label='B').opts(subcoordinate_y=True)
+
+    overlay = (c1 * c2)
+    RangeXY(source=overlay)
+
+    plot = bokeh_server_renderer.get_plot(overlay)
+
+    p1, p2 = plot.subplots.values()
+
+    assert not (p1.callbacks or p2.callbacks)
```
Comment (hoxbro) [inline holoviews/tests/plotting/bokeh/test_callbacks.py:504]:
```suggestion
    assert not p1.callbacks 
    assert not p2.callbacks
```

It's easier for me to read, and if one of them starts failing, it's also easier to see which.

File: holoviews/plotting/bokeh/links.py [hunk 4]
```diff
@@ -141,25 +141,30 @@ def __init__(self, root_model, link, source_plot, target_plot):
             if axis not in link.axes:
                 continue
 
-            axes[f'{axis}_range'] = target_plot.handles[f'{axis}_range']
+            range_name = f'{axis}_range'
+            if f'subcoordinate_{axis}_range' in target_plot.handles:
+                target_range_name = f'subcoordinate_{range_name}'
+            else:
+                target_range_name = range_name
+            axes[range_name] = target_plot.handles[target_range_name]
```
Comment (hoxbro) [inline holoviews/plotting/bokeh/links.py:149]:
```suggestion
            axes[range_name] = ax = target_plot.handles[target_range_name]
```

When we are already changing things, and then use `ax` for the rest of the code.

**Comment by droumis** [comment]
This is amazing (combined with the hold render changes in https://github.com/holoviz/holoviews/pull/6265. Both for notebook and standalone.

https://github.com/holoviz/holoviews/assets/6613202/9c6b01ee-d38b-4c58-982d-25cc50e6f0b6

File: holoviews/plotting/plot.py [hunk 5]
```diff
@@ -969,6 +969,19 @@ class CallbackPlot:
...
+    @staticmethod
+    def _sources_match(src1, src2):
+        return src1 is src2 or (src1._plot_id is not None and src1._plot_id == src2._plot_id)
+
+    def _matching_plot_type(self, element):
+        """
+        Checks if the plot type matches the element type.
+        """
+        return (
+            (not isinstance(element, CompositeOverlay) or isinstance(self, GenericOverlayPlot) or self.batched) and
```
Comment (hoxbro) [inline holoviews/plotting/plot.py:981]:
```suggestion
            (not isinstance(element, CompositeOverlay) or isinstance(self, GenericOverlayPlot) or self.batched) and
```

I just want you to double-check that this is not missing an extra parentheses.

File: holoviews/plotting/bokeh/links.py [hunk 6]
```diff
@@ -94,31 +95,37 @@ def find_links(cls, root_plot):
...
                 tgt_links = [p for p in potentials if p is not None]
                 if tgt_links:
                     found.append((link, plot, tgt_links[0][0]))
         return found
 
     @classmethod
-    def find_link(cls, plot, link=None):
+    def find_link(cls, plot, link=None, target=False):
         """
         Searches a GenericElementPlot for a Link.
```
Comment (hoxbro) [inline holoviews/plotting/bokeh/links.py:107]: Maybe add a small comment about `target` in the docstring.

**Review by hoxbro — APPROVED** [review]

File: holoviews/plotting/bokeh/links.py (hunk 6 above)
Comment (philippjfr) [reply holoviews/plotting/bokeh/links.py]: Done.

**Pull request merged**